Docs update for new Kustomize setup (#46227)
sourcegraph/sourcegraph: doc/admin/deploy/kubernetes/index.md

# Sourcegraph with Kubernetes

<p class="lead">
Deploying Sourcegraph on Kubernetes is for organizations that need highly scalable and available code search and code navigation. We recommend deploying Sourcegraph on Kubernetes with Kustomize.
</p>

<div class="getting-started">
  <a href="./kustomize" class="btn btn-primary" alt="Configure">
   <span>★ Kustomize</span>
   </br>
   Deploy Sourcegraph with simple kubectl commands
  </a>
  <a href="./helm" class="btn" alt="instance">
   <span>Helm</span>
   </br>
   Deploy Sourcegraph with Helm
  </a>
</div>

<div class="getting-started">
<a class="btn btn-primary text-center" href="#prerequisites">★ Installation</a>
<a class="btn text-center" href="kustomize/configure">Configuration</a>
<a class="btn text-center" href="../instance-size">Instance Sizes</a>
<a class="btn text-center" href="operations">Operations</a>
</div>

> > WARNING: If you are currently on Sourcegraph version 4.5.0 or below, please refer to the [deprecated deployment docs for Kubernetes](../deprecated/index.md).

Reviewer: Wouldn't readers be able to access older versions of sourcegraph docs using the `/@version/` notation?
(e.g., https://docs.sourcegraph.com/@v4.2.0/admin/deploy/kubernetes)

### Prerequisites

* [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) v1.19 or later
   - [Kustomize](https://kustomize.io/) (built into `kubectl` in version >= 1.14)
* A [Kubernetes](https://kubernetes.io/) cluster
   - Minimum Kubernetes version: [v1.19](https://kubernetes.io/blog/2020/08/26/kubernetes-release-1.19-accentuate-the-paw-sitive/)
   - Support for Persistent Volumes (SSDs recommended)
   - You can optionally refer to our [terraform configurations](https://github.com/sourcegraph/tf-k8s-configs) for setting up clusters on:
     - [Amazon Web Services EKS](https://github.com/sourcegraph/tf-k8s-configs/tree/main/aws)
     - [Azure AKS](https://github.com/sourcegraph/tf-k8s-configs/tree/main/azure)
     - [Google Cloud Platform GKE](https://github.com/sourcegraph/tf-k8s-configs/tree/main/gcp)

---

### **Step 1**: Set up a release branch

Set up a release branch from the default branch in your local forked copy of the [deploy-sourcegraph-k8s](https://github.com/sourcegraph/deploy-sourcegraph-k8s) repository.

```bash

> $ git checkout -b release

Reviewer: I didn't see a reference to the `repositories.md` instructions for cloning down the reference repository before this command

```

### **Step 2**: Set up an overlay

> `cd` into the repository and create a new directory `instances/my-sourcegraph` for your Sourcegraph instance.

Reviewer: no need to `cd` into the repo if you are already running git commands in the previous step, right?

Reviewer: also the header is using the term "overlay" while the instructions refer to it as an "instance". I think we should try to use consistent terminology to avoid confusion

Reviewer: one last note, it might read more clearly if we refer to the new directory as `instances/<my-sourcegraph-instance-name>` or something like that nudge the reader to set a unique name for their instance instead of using the literal value `my-sourcegraph`, and then use the code block to provide an example:

`mkdir instance/acme-corp-sourcegraph-prod`

Author: fixed

```bash

> $ mkdir instance/my-sourcegraph

Reviewer: minor typo, should be instances (with an **s**)

> $ mkdir instance/my-sourcegraph

Reviewer: ```suggestion
$ mkdir instances/my-sourcegraph
```

```

> ### **Step 3**: Set up a configuration file

Reviewer: ```suggestion
### **Step 3**: Initialize a kustomization file for the instance
```

Copy [instances/kustomization.template.yaml](kustomize/index.md#template) to the `instances/my-sourcegraph` subdirectory as `kustomization.yaml`.

```bash

> $ cp instances/kustomization.template.yaml instances/my-sourcegraph/kustomization.yaml

Reviewer: ```suggestion
$ cp instances/template/kustomization.template.yaml instances/my-sourcegraph/kustomization.yaml
```

```

The new `kustomization.yaml` file will be used to configure your deployment.

### **Step 4**: Set namespace

> Replace `sourcegraph` with a namespace that matches an existing namespace in your cluster. 

Reviewer: ```suggestion
By default, the provided kustomization.yaml template deploys sourcegraph into the `sourcegraph` namespace. If you intend to deploy sourcegraph into a different namespace, replace `sourcegraph` with the name of the existing namespace in your cluster.
```
We could also add a note that the namespace needs to exist/ be created prior to running `kustomize apply`, but that should be obvious for the intended audience of this guide.

Set it to `default` to deploy to the default namespace.

  ```yaml
  # instances/my-sourcegraph/kustomization.yaml
  namespace: sourcegraph
  ```

### **Step 5**: Set storage class

> Sourcegraph requires a storage class that supports SSDs for proper storage and instance performance.

Reviewer: FYI this is contradictory with the note in the prereqs above ("SSDs _recommended_"). I think we should clarify that SSD is _strongly recommended_ for non-production environments, but is a hard requirement for any production deployments of sourcegraph. Or if we want to say it's a requirement for _any_ sourcegraph deployment, we should update the language in the prerequisites to avoid confusion.

Author: Update language to state that SSD is recommended and not required

Here are the options to set the storage class for your instance:

#### Option 1: Create a new storage class

If you have the ability to create a new storage class, choose a storage-class component that corresponds to your cluster provider:

  ```yaml
  # instances/my-sourcegraph/kustomization.yaml
  components:
    - ../../components/storage-class/aws
    - ../../components/storage-class/azure
    - ../../components/storage-class/gke
  ```

Please refer to our [configurations guide](kustomize/configure.md) for a complete list of available storage class components and other specific configuration options.

> #### Option 2: Use an existing storage class

Reviewer: I personally feel like that the default recommendation (i.e., "Option 1") should be to use an existing SSD storage class in the cluster if one exists, and only guide the user to create a new storage class if there isn't one available in the cluster already

Author: Discussed with beyang and we believe recommending users to create storage class to make sure SSDs are provisioned should be the default, but we can revisit this later if that changes

If you'd like to use an exisiting storage class:

1. Include the `storage-class/update-class-name` component under the components list
2. Input the storage class name by setting the value for `STORAGECLASS_NAME` under the configMapGenerator section
   
For example, set `STORAGECLASS_NAME=sourcegraph` if `sourcegraph` is the name of an existing storage class:

  ```yaml
  # instances/my-sourcegraph/kustomization.yaml
  components:
    # Update storageClassName to the STORAGECLASS_NAME value set below
    - ../../components/storage-class/update-class-name
  # ...
  configMapGenerator:
  - name: sourcegraph-kustomize-env
    behavior: merge
    literals:
      - STORAGECLASS_NAME=sourcegraph # Set STORAGECLASS_NAME value to 'sourcegraph'
  ```

#### Option 3: Use default storage class

Skip this step to use the default storage class without SSD support for non-prod environment; however, you must recreate the cluster with SSDs configured for a production environment later.

### **Step 6**: Build manifests with Kustomize

Generate a new set of manifests using the configuration applied to the `my-sourcegraph` subdirectory without applying to the cluster.

  ```bash

>   $ kubectl kustomize instances/my-sourcegraph -o cluster.yaml

Reviewer: ```suggestion
  $ kubectl kustomize instances/<my-sourcegraph-instance-name> -o <my-sourcegraph-instance-name>.yaml
```
We should nudge the reader to set a specific output file name in this command versus using a static value, since repeated executions of this command for different instances will overwrite this file (which is created in the root directory). Also I personally find it confusing to call the output file `cluster.yaml` since we're not creating a cluster with this manifest, but instead are deploying _into_ a (k8s) cluster.

  ```

### **Step 7**: Review manifests

> Review the generated manifests to ensure they match your intended configuration.

Reviewer: ```suggestion
Review the generated manifest to ensure that it matches your intended configuration.
```

Reviewer: I notice that we use plural "manifests" in a few other places. I think we should consistently refer to a singular manifest, since the guide only generates/ applies a single k8s yaml file

Author: The output file contains multiple manifests so I think `manifests` is the right word to use here?

  ```bash
  $ less cluster.yaml
  ```

### **Step 8**: Deploy the generated manifests

Apply the manifests from the ouput file `cluster.yaml` to your cluster:

  ```bash
  $ kubectl apply --prune -l deploy=sourcegraph -f cluster.yaml
  ```

### **Step 9**: Monitor the deployment

Monitor the deployment status to ensure all components are running properly.

  ```bash
  $ kubectl get pods -A -o wide --watch
  ```

### **Step 10**: Access Sourcegraph in Browser

> To access Sourcegraph in a web browser, you must first forward the remote port temporarily:

Reviewer: ```suggestion
To verify that the deployment was successful, port-forward the frontend pod with the following command:
```

  ```bash
  $ kubectl port-forward svc/sourcegraph-frontend 3080:30080
  ```

Then access your new Sourcegraph instance at http://localhost:3080 to proceed to the site-admin setup step.

  ```bash
  $ open http://localhost:3080
  ```

---

## Configure

After the initial deployment, additional configuration might be required for Sourcegraph to customize your deployment to suit your specific needs.

Common configurations that are strongly recommended for all Sourcegraph deployments:

> - [Add monitoring](kustomize/configure.md#monitoring-stack)

Reviewer: ```suggestion
- [Enable the Sourcegraph monitoring stack](kustomize/configure.md#monitoring-stack)
```

> - [Allocate resources based on your instance size](kustomize/configure.md#instance-size-based-resources) (refer to our [instance size chart](../instance-size.md))

Reviewer: I'd suggest saying "Adjust resource allocations" instead, since we technically already allocated resources with the initial deployment
